File: it-agents/skills/code-architect/references/code-examples/cpp/Builder.cpp
// Builder — construct complex objects step by step.
#include <iostream>
#include <memory>
#include <string>
#include <vector>

// The complex product.
struct Assembly {
    std::string frame;
    std::vector<std::string> modules;
    void show() const {
        std::cout << "Assembly[frame=" << frame << ", modules=";
        for (size_t i = 0; i < modules.size(); ++i)
            std::cout << modules[i] << (i + 1 < modules.size() ? "," : "");
        std::cout << "]\n";
    }
};

// Builder interface.
class AssemblyBuilder {
public:
    virtual ~AssemblyBuilder() = default;
    virtual void reset() = 0;
    virtual void setFrame() = 0;
    virtual void addModule() = 0;
    virtual Assembly result() = 0;
};

class StandardBuilder : public AssemblyBuilder {
    Assembly product;
public:
    StandardBuilder() { reset(); }
    void reset() override { product = Assembly{}; }
    void setFrame() override { product.frame = "steel"; }
    void addModule() override { product.modules.push_back("standard-module"); }
    Assembly result() override {
        Assembly out = std::move(product);
        reset();
        return out;
    }
};

// Director encodes a known construction sequence.
class Director {
public:
    Assembly buildMinimal(AssemblyBuilder& b) {
        b.reset();
        b.setFrame();
        return b.result();
    }
    Assembly buildFull(AssemblyBuilder& b) {
        b.reset();
        b.setFrame();
        b.addModule();
        b.addModule();
        return b.result();
    }
};

int main() {
    StandardBuilder builder;
    Director director;
    director.buildMinimal(builder).show();
    director.buildFull(builder).show();
    return 0;
}

File: skills/code-architect/references/code-examples/csharp/Builder.cs
// Builder — construct complex objects step by step.
using System;
using System.Collections.Generic;
using System.Text;

namespace Patterns.Builder
{
    sealed class Product
    {
        public List<string> Parts { get; } = new();
        public string Report()
        {
            var sb = new StringBuilder("Product{");
            sb.Append(string.Join(", ", Parts));
            return sb.Append('}').ToString();
        }
    }

    interface IBuilder
    {
        IBuilder Reset();
        IBuilder AddPartA();
        IBuilder AddPartB();
        IBuilder AddPartC();
        Product Build();
    }

    sealed class ConcreteBuilder : IBuilder
    {
        private Product _product = new();
        public IBuilder Reset() { _product = new Product(); return this; }
        public IBuilder AddPartA() { _product.Parts.Add("PartA"); return this; }
        public IBuilder AddPartB() { _product.Parts.Add("PartB"); return this; }
        public IBuilder AddPartC() { _product.Parts.Add("PartC"); return this; }
        public Product Build() { Product result = _product; Reset(); return result; }
    }

    // Director encapsulates known construction recipes.
    sealed class Director
    {
        public Product BuildMinimal(IBuilder b) => b.Reset().AddPartA().Build();
        public Product BuildFull(IBuilder b) => b.Reset().AddPartA().AddPartB().AddPartC().Build();
    }

    static class Demo
    {
        public static void Run()
        {
            var builder = new ConcreteBuilder();
            var director = new Director();
            Console.WriteLine(director.BuildMinimal(builder).Report());
            Console.WriteLine(director.BuildFull(builder).Report());
            // Custom assembly without the director:
            Console.WriteLine(builder.Reset().AddPartB().AddPartC().Build().Report());
        }

        public static void Main() => Run();
    }
}
